### DISCUSSION TITLE
二次验证失败

### DISCUSSION THREAD

**Discussion by Tan-Pchan** [hhyo/Archery#2198] [Q&A]
求助
版本： 1.9.1
问题： 从1.8.5升级到1.9.1后，用户配置了Google验证方式，登录后二次验证选择项为空，auth_types 为空
![image](https://github.com/hhyo/Archery/assets/68987741/0aba2259-e01b-4153-b33e-981dd6c11ee3)
![image](https://github.com/hhyo/Archery/assets/68987741/762041d8-c721-4ad1-ab85-24b63fef82be)

**Comment by nick2wang** [comment]
升级时是否执行了变更sql？
https://github.com/hhyo/Archery/blob/1ef4374a3af4ba56c5103eb5fe0542e7f17e139b/src/init_sql/v1.9.0.sql#L4-L19

**Comment by Tan-Pchan** [comment]
有执行的。以下为表结构
| 2fa_config | CREATE TABLE `2fa_config` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `username` varchar(200) NOT NULL,
  `auth_type` varchar(128) NOT NULL,
  `phone` varchar(64) DEFAULT '',
  `secret_key` varchar(256) DEFAULT NULL,
  `user_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `2fa_config_user_id_auth_type_303171fc_uniq` (`user_id`,`auth_type`),
  KEY `2fa_config_user_id_a0d1d7c2` (`user_id`),
  CONSTRAINT `2fa_config_user_id_new_fk_sql_users_id` FOREIGN KEY (`user_id`) REFERENCES `sql_users` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=27 DEFAULT CHARSET=utf8mb4 |

**Reply by nick2wang** [reply]
看一下2fa_config表里该用户的配置

**Reply by Tan-Pchan** [reply]
username auth_type secret_key这三个字段的值都是加密的。

**Reply by nick2wang** [reply]
在管理页面看下
http://archery-host/admin/sql/twofactorauthconfig/

**Reply by Tan-Pchan** [reply]
后台管理界面，是正常的用户信息
![image](https://github.com/hhyo/Archery/assets/68987741/4ed9e13c-01fb-4541-afe7-7eca48a4c4d7)

**Reply by Tan-Pchan** [reply]
这是我用旧版1.8.5的管理后台截图的。1.9.1版本登不进去

**Reply by nick2wang** [reply]
1.9.1版本上执行看下有没有结果
``` sh
python manage.py shell
```
``` python
from sql.models import TwoFactorAuthConfig
twofa_enabled = TwoFactorAuthConfig.objects.filter(username="你的用户名")
auth_types = [twofa.auth_type for twofa in twofa_enabled]
print(auth_types)
```

**Reply by Tan-Pchan** [reply]
是空数组
[]

**Reply by nick2wang** [reply]
这个账号改过用户名么？

**Reply by Tan-Pchan** [reply]
没有的。试了登录两个账号都一样的结果。

**Reply by Tan-Pchan** [reply]
二次验证写死Google验证码方式后，登录报错。
![image](https://github.com/hhyo/Archery/assets/68987741/88666e3a-3306-409e-8c66-462faa701872)
增加输出打印，user打印的是显示中文名，非user用户对象
![image](https://github.com/hhyo/Archery/assets/68987741/6319786f-3af6-420f-bf7e-6bb33a2b6dea)

**Reply by nick2wang** [reply]
登录验证账号密码时把这个结果打印出来看下
https://github.com/hhyo/Archery/blob/813a16e37fa0ad63eaa3de80ceed11522d41e6a7/common/auth.py#L121

**Reply by Tan-Pchan** [reply]
![image](https://github.com/hhyo/Archery/assets/68987741/3c9f96e4-7c09-4b1f-9111-573b7fd1fdd8)

**Reply by nick2wang** [reply]
应该是2fa_config的用户名和user_id不一致导致，登录验证时第一次查询2fa_config是用的user object（即id），第二次校验2fa页面查询用的是username

**Reply by Tan-Pchan** [reply]
刚试了1.9.0版本也是一样的情况。
那请问现在这种情况要怎么修复？

**Reply by nick2wang** [reply]
确认下这条2fa_config记录的username和user_id是不是正确对应的

**Reply by Tan-Pchan** [reply]
2fa_config表中的username是加密的
![image](https://github.com/hhyo/Archery/assets/68987741/91447edf-244e-481b-a486-eb42c7b565e8)

**Reply by nick2wang** [reply]
管理页面可以查看

**Reply by Tan-Pchan** [reply]
![image](https://github.com/hhyo/Archery/assets/68987741/0198f74c-5911-473e-b886-1449e6277eb2)
有对应上的

**Reply by nick2wang** [reply]
那就有点奇怪了，你把浏览器缓存和cookie都清理再试下

**Reply by Tan-Pchan** [reply]
清理后，重登，二次验证时依然报错：sql.models.TwoFactorAuthConfig.DoesNotExist: TwoFactorAuthConfig matching query does not exist.

**Reply by nick2wang** [reply]
把这条记录删掉，重新配置2fa看有没有问题
